src/util/templates/ChainedTemplateEngine.ts

```diff
@@ -0,0 +1,37 @@
+import type { Template, TemplateEngine } from './TemplateEngine';
+import Dict = NodeJS.Dict;
+
+/**
+ * Calls the given array of {@link TemplateEngine}s in the order they appear,
+ * feeding the output of one into the input of the next.
+ *
+ * The first engine will be called with the provided contents and template parameters.
+ * All subsequent engines will be called with no template parameter and contents being `{ body }`,
+ * with `body` the output of the previous engine.
+ */
+export class ChainedTemplateEngine<T extends Dict<any> = Dict<any>> implements TemplateEngine<T> {
+  private readonly contentsEngine: TemplateEngine<T>;
+  private readonly bodyEngines: TemplateEngine<{ body: string }>[];
+
+  /**
+   * @param engines - Engines will be executed in the same order as the array.
+   *                  Actual expected type is `[ TemplateEngine<T>, ...TemplateEngine<{ body: string }>]`
+   */
+  public constructor(engines: TemplateEngine[]) {
+    if (engines.length === 0) {
+      throw new Error('At least 1 engine needs to be provided.');
+    }
+    this.contentsEngine = engines[0];
+    this.bodyEngines = engines.slice(1);
+  }
+
+  public async render(contents: T): Promise<string>;
+  public async render<TCustom = T>(contents: TCustom, template: Template): Promise<string>;
+  public async render<TCustom = T>(contents: TCustom, template?: Template): Promise<string> {
+    let body = await this.contentsEngine.render(contents, template!);
```

Comment:
`!` is needed because of the strict typings on the render function not allowing you to call it with an optional template 😅

Reply:
The only way to not need it would be to revert the typing changes on the `TemplateEngine` interface and just have 1 function with the optional template again.

Reply:
Right, no that's fine.